### PR TITLE
feat: add mocking errors to ic-http

### PR DESCRIPTION
**Pull request by maksymar**
This PR adds the possibility of mocking canister rejection errors to ic-http.

Currently it's only possible to mock responses when calling `ic_http::http_request(...).await`, but with this change it'd be possible to mock canister rejection errors.

Additionally simplify example_canister code, add more tests for errors.

### CODE REVIEW

File: ic-http/src/mock.rs [hunk 1]
```diff
@@ -8,7 +8,8 @@ use std::time::Duration;
 #[derive(Clone)]
 pub(crate) struct Mock {
     pub(crate) request: CanisterHttpRequestArgument,
-    response: HttpResponse,
+    response: Option<HttpResponse>,
+    error: Option<(RejectionCode, String)>,
```
Comment (dsarlis) [inline ic-http/src/mock.rs:12]: How about we have a single field `result` which would be `Option<Result<HttpResponse, (RejectionCode, String)>>`. The reason is that this way it's more clear that it can be either one or another and it's enforced through the type system.

File: ic-http/src/transform.rs [hunk 2]
```diff
@@ -83,4 +84,48 @@ mod test {
         // Assert.
         assert_eq!(names, vec!["transform_function_1", "transform_function_2"]);
     }
+
+    /// Transform function which is intentionally creates a new request
+    /// with itself as a transform function.
```
Comment (dsarlis) [inline ic-http/src/transform.rs:89]: This sentence is a bit hard to parse. Maybe rewrite it a bit? Something like:
```
Transform function which intentionally creates a new request passing itself as the target transform function.
```

File: ic-http/src/transform.rs [hunk 3]
```diff
@@ -83,4 +84,48 @@ mod test {
...
+        arg.response
+    }
+
+    /// Creates a request with a transform function which overwrites itself.
+    fn create_request_with_transform() -> CanisterHttpRequestArgument {
+        crate::request::create_request()
+            .url("https://www.example.com")
+            .transform_func(transform_function_with_overwrite, vec![])
+            .build()
+    }
```
Comment (dsarlis) [inline ic-http/src/transform.rs:104]: Actually, reading this function, how about we move `args.response` from above inside here directly? Not sure we gain much by having the extra `transform_function_with_overwrite`.

**Review by dsarlis — APPROVED** [review]
Looks good. Left some minor comments which I don't think are blocking.

**Comment by dsarlis** [comment]
> Looks good. Left some minor comments which I don't think are blocking.

Ah I didn't realise you had set it to be merged automatically on approval. Feel free to address in a follow-up if you agree with the suggestions.

File: ic-http/src/mock.rs (hunk 1 above)
Comment (maksymar) [reply ic-http/src/mock.rs:12]: Done in #185, good idea, thank you!

File: ic-http/src/transform.rs (hunk 3 above)
Comment (maksymar) [reply ic-http/src/transform.rs:104]: This has to be a separate function with a name so it can be stored in a thread-local storage by it's name.

File: ic-http/src/transform.rs (hunk 2 above)
Comment (maksymar) [reply ic-http/src/transform.rs:89]: Done in #185.

**Comment by maksymar** [comment]
> > Looks good. Left some minor comments which I don't think are blocking.
> 
> Ah I didn't realise you had set it to be merged automatically on approval. Feel free to address in a follow-up if you agree with the suggestions.

I was playing around with this button and forgot to turn it off, sorry about that!
Thank you for the review, I addressed your comments in #185, PTAL.
